Add BatteryPool implementation for aggregating battery-inverter data

ela-kotulska-frequenz opened the pull request:
BatteryPool implementation for aggregating battery-inverter metrics into higher level metrics

ela-kotulska-frequenz committed:
Add maxsize argument when subscribing for the component data

Default maxsize=50 is not working in case when user is interested
in the latest component data only.
maxsize has default constant argument specified outsided.
That way this constanst can be reused and this PR is not
breaking change.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Add option to get batteries status via PowerDistributingActor

Add batteries_status_sender to the PowerDistributingActor.
This sender will send batteries status when it change.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Add metric calcuator for BatteryPool

MetricCalculator defines what components and metrics are needed
to compute high level metrics and how to calculate it.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Add metric fetcher for the battery pool

Metric fetcher defines how to subscribe for the component metrics
and receive the component metrics data.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Add BatterPool interface

BatteryPool is and interface for user to receive the aggregated
battery-inverter data.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Add example how to use battery pool.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

shsms commented on src/frequenz/sdk/microgrid/client/_client.py:
```diff
@@ -80,6 +80,7 @@ async def connections(
     async def meter_data(
         self,
         component_id: int,
+        maxsize: Optional[int] = None,
```
I feel like making it optional means that there is no max size,  like it will never overflow.  And it makes the default value opaque, and it is not a one step to find the correct default value.

What do you think about making it a constant `RECEIVER_MAX_SIZE = 50` in this file instead?  I'll do that for the formula receiver too.  That way we have the flexibility to change the default if necessary, just for this case.

Maybe we really should just export a constant from the channels repo,  but we can make an issue for that.

ela-kotulska-frequenz replied:
Done

shsms commented on src/frequenz/sdk/actor/power_distributing/_battery_pool_status.py:
```diff
@@ -121,7 +124,7 @@ def __init__(
         )
 
-        self._task = asyncio.create_task(self._run())
+        self._task = asyncio.create_task(self._run(battery_status_sender))
```
why not an instance variable instead?  then you don't have to document it 3 times,  or pass it around, to two more functions.

ela-kotulska-frequenz replied:
Because it is not needed as an instance variable. 
There is already lots of instance variables.

leandro-lucarella-frequenz replied:
For what's worth, I agree with Sahas that it would make sense to make it an instance variable, it seems like something very inherent to the class. If at some point you would want to update the sender, it won't be possible if you pass it around. But for me is minor and I'm fine to leave it as it is too if you feel it shouldn't be an instance variable, in the extremely rare event that we want to give the option to replace the sender in the future we can change it then.

ela-kotulska-frequenz commented:
`active_power` is calculated for working batteries only. It should be computed for all batteries instead.
`active_power` is calculated using logical meter, now.
I will remove it from this PR and add it in the later one. It will use formula engines. (like in ev pool)

leandro-lucarella-frequenz commented on examples/battery_pool.py:
```diff
@@ -0,0 +1,114 @@
+PORT = 61060
+
+
+def create_battery_pool() -> BatteryPool:
```
This is a good summary of what needs to be put in a SDK initialization function provided by the SDK :D

ela-kotulska-frequenz replied:
Yes :D How soon we will have time for this :)

leandro-lucarella-frequenz commented on src/frequenz/sdk/actor/power_distributing/_battery_pool_status.py:
```diff
@@ -72,13 +72,16 @@ class BatteryPoolStatus:
         Args:
             battery_ids: set of batteries ids that should be stored in pool.
+            battery_status_sender: Channel for sending information which batteries are
+                working.
```
This is not really a channel... Maybe:

```suggestion
            battery_status_sender: The sender used for sending the status of the
                batteries in the pool.
```

ela-kotulska-frequenz replied:
Done

leandro-lucarella-frequenz commented on src/frequenz/sdk/microgrid/component/battery_utils.py:
```diff
@@ -0,0 +1,60 @@
+    return inverter_id
+
+
+def battery_inverter_mapping(batteries: Set[int]) -> Dict[int, int]:
```
If we emit the log message in `find_adjacent()` this function is reduced to:

```py
bat_inv_map = {bat_id: graph.find_adjacent(bat_id, ComponentCategory.INVERTER)
    for battery_id in batteries}:
```

I'm not sure we need a function for this, unless it is something that will be used very often in many places.

If we want to keep it as a function, I think it could also be generalized as a `ComponentGraph` method, so users that want exactly what this function does could do (I would default to raise an exception if a mapping is missing, as is the safest choice):

```py
bat_inv_map = component_graph.map_adjacents(battery_ids, ComponentCategory.INVERTER,
    ignore_missing=True)
```

ela-kotulska-frequenz replied:
Component graph has already method to find successors and predecessors.
They are well defined and I won't add unknown `adjacent` there.
This method is highly related with battery because we are sure inverter is direct predecessor of the battery.

I moved battery_inverter_mapping to the battery_pool module. It is now private.

leandro-lucarella-frequenz commented on src/frequenz/sdk/microgrid/component/battery_utils.py:
```diff
@@ -0,0 +1,60 @@
+_logger = logging.getLogger(__name__)
+
+
+def find_adjacent_inverter_id(battery_id: int) -> Optional[int]:
```
I wonder if this doesn't belong to the `ComponentGraph`. It could be generalized to find adjacent inverters to any component (maybe it can be parametrized to find any type of component adjacent to other type of componet). Not sure if we have use case for something this generic yet but it sounds like it might be a good utility to have?

Like:

```py
inverters = component_graph.find_adjacent(battery_id, ComponentCategory.INVERTER)
```

ela-kotulska-frequenz replied:
Actually component graph has now methods to find successors/predecessors.
And I prefer successor/predecessor instead of `adjacent` because it is more error resilient.

leandro-lucarella-frequenz commented on src/frequenz/sdk/microgrid/component/battery_utils.py:
```diff
@@ -0,0 +1,60 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Util functions for battery."""
```
This module sounds a bit weird for me, it feels like if it is something that it is used only for batteries and frequently by SDK users, it should be in the `BatteryPool` somehow (as we want that to be our interface to batteries, right)?

Otherwise it seems like they could be utility functions from the component graph. Se the comments on the specific functions.

ela-kotulska-frequenz replied:
Module removed. 
Method moved to the BatteryPool.

leandro-lucarella-frequenz commented on src/frequenz/sdk/microgrid/component/battery_utils.py:
```diff
@@ -0,0 +1,60 @@
+            None,
+        )
+    except KeyError as err:
+        _logger.warning(err)
```
Do we really want to log an error each time someone tries to find an inverter for a battery and it's not there? This error is also duplicated when the `battery_inverter_mapping()` function is used.

Also, if we keep it, I would write a more descriptive error instead of printing the `err` object raw.

BTW, is a microgrid in a valid state if it has batteries without inverters? Because if it's not maybe this should be an error instead.

ela-kotulska-frequenz replied:
> Do we really want to log an error each time someone tries to find an inverter for a battery and it's not there? This error is also duplicated when the battery_inverter_mapping() function is used.

This error tells there is no component with this id in component graph. So yes it would be nice to print it :D

> Also, if we keep it, I would write a more descriptive error instead of printing the err object raw.

It is already descriptive enough in the ComponentGraph.

> BTW, is a microgrid in a valid state if it has batteries without inverters? Because if it's not maybe this should be an error instead.

It is valid state to have inverter without battery. Few months ago we added it.
I think component graph allows for battery without inverter, so it is valid state, too. (I can check it if necessary) 
Anyway these requirements can change in the future, so I prefer to check it here. It is one-time check and cost me nothing.

leandro-lucarella-frequenz replied:
Yeah, what I am more concerned about is *false-positive* errors. I.e. if it is normal for a battery not to have an inverter, when we shouldn't print a warning or an error, because then we get the error logs littered with stuff that it's just normal and then it is really easy to miss real errors in the noise.

ela-kotulska-frequenz replied:
True, thanks for noting that!
I will change it to `info` log :)

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/__init__.py:
```diff
@@ -0,0 +1,10 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Module to aggregate data from pool of the batteries."""
```
I would skip the `Module to`:

```suggestion
"""Aggregates data from a pool of batteries."""
```
or

```suggestion
"""Manage a pool of batteries."""
```

If we'll provide also controlling functions with the battery pool (still not sure if charge/discharge will be done via the battery pool or power manager).

ela-kotulska-frequenz replied:
Done, changed to `Manage a pool of batteries.`
Because it will probably charge/discharge batteries in the future.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_component_metric_fetcher.py:
```diff
@@ -0,0 +1,305 @@
+    def __init__(
+        self,
+        component_id: int,
+        metrics: List[ComponentMetricId],
```
Super minor, but if you are copying anyway, maybe you can require just a `Iterable[ComponentMetricId]`, so you can pass iterators or generators too.

ela-kotulska-frequenz replied:
Done

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_component_metric_fetcher.py:
```diff
@@ -0,0 +1,305 @@
+        if any(map(lambda mid: mid not in _BatteryDataMethods, metrics)):
+            raise ValueError(
+                "Some battery metrics are not supported, supported: "
+                f"{str(_BatteryDataMethods.keys())}, requested: {str(metrics)}"
```
```suggestion
                f"Some battery metrics are not supported: {list(set(metrics) - _BatteryDataMethods.keys())}"
                f" (supported metrics are: {list(_BatteryDataMethods.keys())}"
```

ela-kotulska-frequenz replied:
I simplified whole log message

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_component_metric_fetcher.py:
```diff
@@ -0,0 +1,305 @@
+        if any(map(lambda mid: mid not in _InverterDataMethods, metrics)):
+            raise ValueError(
+                "Some battery metrics are not supported, supported: "
+                f"{str(_InverterDataMethods.keys())}, requested: {str(metrics)}"
```
Same

ela-kotulska-frequenz replied:
I simplified whole log message

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_component_metric_fetcher.py:
```diff
@@ -0,0 +1,305 @@
+
+
+@dataclass
+class ResamplerArgs:
```
I was a bit confused by this class, because is not really arguments for the resampler. Is more like a bag of stuff that is used in connection to make requests to the resampler actor. Any reason against just passing these separately to the `ResampledComponentMetricFetcher` instead? They seem to be used only in `_subscribe_for_metric()` anyway.

In the long term, I think we need to simplify the interface to subscribe to the resampler, there is a lot of boilerplate going on here.

ela-kotulska-frequenz replied:
Removed.
Resampler was needed for `active power` only , but we will use FormulaEngine for `active power`.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_component_metric_fetcher.py:
```diff
@@ -0,0 +1,305 @@
+    """Channel registry instance to get the receiver for the resampling channel."""
+
+
+class ResampledComponentMetricFetcher(ComponentMetricFetcher):
```
This is somehow analogous to the `LogicalMeter`, right?

ela-kotulska-frequenz replied:
Yes, but was formula was computed for  working  components only.
In the end it turns out we need to compute active_power for all batteries in pool. (independently if they are working or not).
So I removed this class and I will use FormulaEngine for active power in future (next PR)

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_methods.py:
```diff
@@ -0,0 +1,413 @@
+_MIN_TIMESTAMP = datetime.min.replace(tzinfo=timezone.utc)
+
+
+class BaseMethod(ABC, Generic[T]):
```
Maybe `AggregateMethod` or `SendMethod` whatever this is? Without reading the code I have no idea if this is a method for sending, aggregating or what :)

ela-kotulska-frequenz replied:
Done : changed to `AggregateMethod`

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py:
```diff
@@ -0,0 +1,475 @@
+        Returns:
+            Name of the calculator
+        """
+        return "Soc_Capacity"
```
If this is only for debugging, maybe you can move it to `MetricCalculator` and use an implementation like `return cls.__name__`.

ela-kotulska-frequenz replied:
No, I use is outside in battery_pool module to create unique name for the active methods

leandro-lucarella-frequenz replied:
OK, couldn't this just be `Capacity` now?

ela-kotulska-frequenz replied:
Done, thanks!

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py:
```diff
@@ -0,0 +1,475 @@
+
+# Soc and capacity are different kind of metrics.
+# It makes sens to separate them with `_`.
+class SoC_Capacity_Calculator(
```
Why underscore in class names here?

ela-kotulska-frequenz replied:
I separated it as we discussed on the meeting

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_result_types.py:
```diff
@@ -0,0 +1,181 @@
+# Soc and capacity are different kind of metrics.
+# It makes sens to separate them with `_`.
+@dataclass
+class SoC_Capacity_Metrics:  # pylint: disable=invalid-name
```
I don't get this. Why can't we use `SoCCapacityMetrics`?

I'd prefer to avoid at all cost diverging from the class naming convention.

If this refers to a group of different metrics, and this is basically all the metrics a battery pool will have, can't it be `BatteryPoolMetrics`? If is not for the whole thing maybe `SocCapacityCombinedMetrics`?

ela-kotulska-frequenz replied:
I separated it as we discussed on the meeting

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_result_types.py:
```diff
@@ -0,0 +1,181 @@
+# Soc and capacity are different kind of metrics.
+# It makes sens to separate them with `_`.
+@dataclass
+class SoC_Capacity:  # pylint: disable=invalid-name
```
Same here. Maybe `SocCapacityCombined`?

ela-kotulska-frequenz replied:
I separated it as we discussed on the meeting

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/battery_pool.py:
```diff
@@ -0,0 +1,142 @@
+    BatterPool provided methods for fetching different battery and inverter values
+    from microgrid and aggregate then in a high level metrics.
+    Depending from the metrics the results are send in two ways:
+        * when the metric change
+        * with the resampling frequency (specified in the resampler actor outside)
```
I think this won't render nicely if it has the extra indent, also a new line makes it more clear.

```suggestion

    * when the metric change
    * with the resampling frequency (specified in the resampler actor outside)
```

ela-kotulska-frequenz replied:
Removed. We have only metrics sending on update in this PR.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py:
```diff
@@ -0,0 +1,475 @@
+
+        if len(batteries) != len(self._bat_inv_map):
+            _logger.warning(
+                "Some batteries hasn't adjacent inverter. Use %s out of %s",
```
This is a 3rd warning for the missing inverter, you **really hate** missing inverters :laughing:

ela-kotulska-frequenz replied:
I don't hate them :D  

It is just still unknown how much inverters battery can have.
First we could have ONLY one inverter per one battery.
Now it can have ZERO or ONE inverter per one battery
In the future we plan to have [0-inf] inverters per one battery. 

But you are right . There are to much logs.  I simplified them

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/battery_pool.py:
```diff
@@ -0,0 +1,142 @@
+            for item in self._active_methods.values():
+                item.update_working_batteries(self._working_batteries)
+
+    async def soc_capacity(
```
This goes in general for the common interface for stuff returning new receivers (i.e. logical meter too). I think we should be more explicit about a **new** receiver being created, otherwise people might think they can just call `soc_capacity()` whenever they need a receiver. It will make the method names a bit uglier, but I think is more clear.

1. `p.new_soc_capacity_receiver()`?
2. Maybe return a channel(or wrapper) instead, so users do `p.soc_capacity.new_receiver()`?
3. `p.new_receiver(SOC_CAPACITY)`?

Can be left for a future PR, but I would like to address it soon, before we keep adding methods like these.

/cc @matthias-wende-frequenz

leandro-lucarella-frequenz replied:
I was thinking that with approach 2 we could add different ways to retrieve the data:

```py
current = await p.soc_capacity.current_value  # Queries the microgrid API (if it's even possible :D)
last = p.soc_capacity.last_value  # If we have a channel getting this, we could peek into it
resampled_receiver = p.soc_capacity.new_receiver()
change_receiver = p.soc_capacity.new_changes_receiver()
```

This would also remove the inconsistent interface we have right now, where `soc_capacity()` could be sending messages to the channel on every resample period, or on change, is not explicit, and it would be inconsistent for different metrics, which makes me uneasy.

We could also have `new_receiver(only_changes=True)` for example, but if we don't want to support getting just a resampled receiver, we need to assert `only_changes` is `True` at runtime, and we can´t use the type system to help us find this at "compile-time".

leandro-lucarella-frequenz replied:
https://github.com/frequenz-floss/frequenz-sdk-python/discussions/215

leandro-lucarella-frequenz reviewed (commented):
I did a "quick" (with big fat quotes) look and didn't review all the algorithmic stuff and focused more on structure and such.

I know it is hard for big features, but if this could have been done in smaller PRs, I think I would have made reviewing much easier.

ela-kotulska-frequenz commented:
I have to fix commit messages and add unit tests.
But implementation is ready for the review

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py:
```diff
@@ -0,0 +1,478 @@
+_MIN_TIMESTAMP = datetime.min.replace(tzinfo=timezone.utc)
+
+
+def battery_inverter_mapping(batteries: Iterable[int]) -> Dict[int, int]:
```
Super nitpick, but if this is only used in this module I would make it even *more private*.
```suggestion
def _battery_inverter_mapping(batteries: Iterable[int]) -> Dict[int, int]:
```

Or just move it as a private method of `PowerBoundsCalculator` to make the scope even more limited if it's just an utility function used only by that class.

ela-kotulska-frequenz replied:
Done

leandro-lucarella-frequenz replied:
I can't see that, in the current state I see the name is still public but now it's being used somewhere else (`_methods.py`), so all good anyway. Feel free to resolved this one!

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py:
```diff
@@ -0,0 +1,478 @@
+    for battery_id in batteries:
+        try:
+            predecessors = graph.predecessors(battery_id)
+        except KeyError as err:
+            _logger.error(str(err))
```
```suggestion
        except KeyError:
            _logger.exception("Can't find predecessors for battery with ID %s", battery_id)
```

ela-kotulska-frequenz replied:
_logger.exception will print whole traceback, too. It is ok to us it for logging really important error messages.
This message is not so important and it doesn't need traceback.

leandro-lucarella-frequenz replied:
In my experience it is always nice to have a backtrace in case you need to investigate something a bit more, which is really hard when problems happen in production and you can't reproduce it easily afterwards. I'm concern about thinking that this error could only happen when a very obvious condition happens too, so normally it should be obvious to figure out why this happen, but what if this is not true for some obscure case we didn't foresee?

So my question is, do you **really** want to avoid printing the backtrace? What would be the disadvantage or doing so?

ela-kotulska-frequenz replied:
I remember it was hard for me to read logs from async application where you have traceback for every single (not very important) error message.
At some point people will learn to ignore such traceback in logs.

I would assume logging.exception is for very important exceptions where this traceback is really needed.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_component_metric_fetcher.py:
```diff
@@ -0,0 +1,248 @@
+            # pylint: disable=protected-access
+            if metric not in self._supported_metrics():
+                category = self._component_category()
+                raise ValueError(f"Metric {metric} not supported fro the {category}")
```
Small typo + minor improvement suggestion:

```suggestion
                raise ValueError(f"Metric {metric} not supported for {category}")
```

ela-kotulska-frequenz replied:
Done

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_component_metric_fetcher.py:
```diff
@@ -0,0 +1,248 @@
+                "Channel for component %d was closed. Err: %s",
+                self._component_id,
+                str(err),
+            )
```
```suggestion
        except ChannelClosedError:
            _logger.exception(
                "Channel for component %d was closed.",
                self._component_id,
            )
```

ela-kotulska-frequenz replied:
Done

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py:
```diff
@@ -0,0 +1,478 @@
+
+
+class SoCCalculator(MetricCalculator[SoCMetrics]):
+    """Define how to calculate Capacity metrics."""
```
Shouldn't this be `SoC` instead of `Capacity`?

ela-kotulska-frequenz replied:
Done

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py:
```diff
@@ -0,0 +1,478 @@
+        Returns:
+            Name of the calculator
+        """
+        return "Soc_Capacity"
```
And this `"SoC"` too?

ela-kotulska-frequenz replied:
Done

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py:
```diff
@@ -0,0 +1,478 @@
+        Returns:
+            Name of the calculator
+        """
+        return "PowerBoundsMetrics"
```
Maybe `"PowerBounds"` without `Metrics` to be consistent with other names?

ela-kotulska-frequenz replied:
Done

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_component_metric_fetcher.py:
```diff
@@ -0,0 +1,248 @@
+
+    @abstractmethod
+    def _component_category(self) -> ComponentCategory:
+        pass
```
Funny that you don't get a complain from `mypy` here. Usually `pass` means that you have an implementation that does nothing, so it should complain about not returning a `float`. For abstract methods you should use `...` instead.
```suggestion
    @abstractmethod
    def _extract_metric(self, data: T, mid: ComponentMetricId) -> float:
        ...

    @abstractmethod
    def _supported_metrics(self) -> Set[ComponentMetricId]:
        ...

    @abstractmethod
    def _component_category(self) -> ComponentCategory:
        ...
```

ela-kotulska-frequenz replied:
Thanks, fixed.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_component_metrics.py:
```diff
@@ -0,0 +1,89 @@
+                "Comparing %s with %s not supported",
+                type(ComponentMetricsData),
+                type(other),
+            )
```
I find this odd, usually in Python it is valid to compare 2 arbitrary types, it will just return `False` if they are not equal. Do we really want to print an error here? I would avoid it unless there is a very good reason.

ela-kotulska-frequenz replied:
Removed

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_methods.py:
```diff
@@ -0,0 +1,265 @@
+    _cached_metrics: Dict[int, ComponentMetricsData]
+
+    _update_task: asyncio.Task[None]
+    _send_task: asyncio.Task[None]
```
Just out of curiosity, is this intentional or was this a `AsyncConstructible` class before and is a left over? I remember us talking in a meeting about the possibility to declare types in the class instead of the `__init__()`, but I'm not sure if we agreed on doing so, one problem was that documentation needs to be in the `__init__()` to render nicely, so I think we should still prefer `__init__()` even if we don't really need to add documentation for private members.

We can still have a block defining types at the beginning of the `__init__()`, for example:

```py
    def __init__(...) -> None:
        _working_batteries: Set[int]
        """The set of batteries that are currently working."""

        _metric_calculator: MetricCalculator[T]
        """The metric calculator used to..."""

        # Codes with assignments goes here.
```

ela-kotulska-frequenz replied:
Yes, it was AsyncConstructible before.
But it is not anymore so I removed it.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_methods.py:
```diff
@@ -0,0 +1,265 @@
+
+    async def stop(self) -> None:
+        """Stop the all pending async task."""
+        await asyncio.gather(
+            *[cancel_and_await(task) for task in self._pending_data_fetchers],
+            return_exceptions=True,
+        )
+        await asyncio.gather(
+            *[cancel_and_await(self._send_task), cancel_and_await(self._update_task)],
+            return_exceptions=True,
+        )
```
So you are just silently swallowing exceptions? Maybe at least print a log if there are any?

In any case all this should be handled by the `BackgroundService` class (if I ever find some time to take care of it :grimacing: ).

ela-kotulska-frequenz replied:
It should not raise exception, because exception is catched in the `cancel_and_await`. 
If for some reason it will raise and exception then something is really really wrong and should not be catched.
So I removed `return_exceptions=True`, thanks

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py:
```diff
@@ -0,0 +1,478 @@
+        """
+
+    @property
+    def batteries(self) -> Set[int]:
```
Warning: This is using `typing.Set` which is analogous to `builtins.set`, which means the user will be able to mutate the set, which is probably not what you want since you made it a property. If you use [`collections.abc.Set`](https://docs.python.org/3.11/library/collections.abc.html#collections.abc.Set) instead, then is an interface for a immutable set, so the user won't be allowed to change it (if they run `mypy` that is, of course at runtime they can do whatever they want :D)

ela-kotulska-frequenz replied:
Done, thanks!

ela-kotulska-frequenz replied:
Unfortunately it doesn't work for python3.8 :( 
```
_____ ERROR collecting tests/timeseries/_battery_pool/test_battery_pool.py _____
tests/timeseries/_battery_pool/test_battery_pool.py:22: in <module>
    from frequenz.sdk.timeseries.battery_pool import (
src/frequenz/sdk/timeseries/battery_pool/__init__.py:7: in <module>
    from .battery_pool import BatteryPool
src/frequenz/sdk/timeseries/battery_pool/battery_pool.py:23: in <module>
    class BatteryPool:
src/frequenz/sdk/timeseries/battery_pool/battery_pool.py:40: in BatteryPool
    batteries_id: Set[int] | None = None,
E   TypeError: 'ABCMeta' object is not subscriptable
```

ela-kotulska-frequenz replied:
I can surround it with ditto mark ( batteries_id: "Set[int]" | None = None, )
But doing it everywhere seems to much.
We can change it when we delete support for python3.8.

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py:
```diff
@@ -0,0 +1,478 @@
+
+    @property
+    @abstractmethod
+    def battery_metrics(self) -> Dict[int, List[ComponentMetricId]]:
```
Maybe for this too, not sure how is it intended to be used, but maybe [`collections.abc.Mapping`](https://docs.python.org/3.11/library/collections.abc.html#collections.abc.Mapping) and [`collections.abc.Sequence`](https://docs.python.org/3.11/library/collections.abc.html#collections.abc.Sequence).

Etc...

ela-kotulska-frequenz replied:
I forgot, about this comment. I will fix it tomorrow.

ela-kotulska-frequenz replied:
Done

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_metric_calculator.py:
```diff
@@ -0,0 +1,478 @@
+            result.bound.upper += capacity * soc_upper_bound
+            result.bound.lower += capacity * soc_lower_bound
+
+        return None if result.timestamp == _MIN_TIMESTAMP else result
```
I wonder how expensive are the calculations above, and if it wouldn't make sense to first do a pass over the timestamps to see if it makes sense to calculate anything at all.

I guess now you are optimizing for the case where everything is working normally and what I propose will be optimizing for the case where data is missing, so maybe is not worth it, but sharing it just in case. :)

ela-kotulska-frequenz replied:
hmm good point, however it is extremely rare that all components of certain type are broken.
It would add extra complexity. And I would (most probably) have to check it anyway because mypy /pylint doesn't see that it is not None. :(

leandro-lucarella-frequenz commented on src/frequenz/sdk/timeseries/battery_pool/_result_types.py:
```diff
@@ -0,0 +1,131 @@
+
+
+@dataclass
+class PowerBoundsMetrics:
```
I wonder if it wouldn't be nicer for the user if this is called `PowerMetrics` instead, since then you have `bound` as memebers so it gets a bit redundant to say `PowerBoundsMetrics.supply_bound` for example. Also not sure if this could ever happen, but maybe in the future we need to add more power related metrics?

ela-kotulska-frequenz replied:
Done

leandro-lucarella-frequenz reviewed (commented):
Most very minor comments. I love how such an amount of complexity gets hidden in a very simple and clean interface :tada:

ela-kotulska-frequenz committed:
Add tool to stream mock component data

In our current MockMicrogrid is component data are hardcoded.
It is impossible to control what component data are send
and to change it for new tests.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Add function to create mock component graph structure

... to pytest utils folder.
The behaviour is copied from the MockMicrogrid.
I moved it to another module to make for readable.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Add temporary sdk interface to pytest utils module

It allows to create all tools in all places and close
them when not needed.
This should be replaced by true sdk interface in the future.

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

ela-kotulska-frequenz committed:
Add unit tests for battery pool

Signed-off-by: ela-kotulska-frequenz <elzbieta.kotulska@frequenz.com>

leandro-lucarella-frequenz reviewed (dismissed):
I think this is good enough to give an approval, it is a quite big PR and there are only very minor things left and I think we can discuss it forever, but it is important to get it in.

For me there is only https://github.com/frequenz-floss/frequenz-sdk-python/pull/205#discussion_r1126277745 left, but I consider it also minor, I just don't see a reason not to go with the traceback and get all the information we can when there is a problem :)

I leave it to you @ela-kotulska-frequenz to decide if to merge it or wait for Sahas to confirm he's fine too.

ela-kotulska-frequenz commented:
Only rebased so I merge without re-aproval

ela-kotulska-frequenz commented:
> to decide if to merge it or wait for Sahas to confirm he's fine too.

We can make fixes when Sahas come back .
Also giving someone so long PR immediately after holiday is to cruel for me :D
